torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3661,8 +3661,20 @@ def _maybe_guard_eq(self, expr: Union["sympy.Eq", "sympy.Ne"], concrete_bool: bo
         # In case of really gnarly expression, we don't blow up
         if len(free) > 5:
             return
-        # NB: prioritize unbacked symints for solving by ordering them last
-        free = sorted(free, key=lambda x: (self.size_hint(x, allow_none=True) or sys.maxsize, x.name), reverse=True)  # type: ignore[attr-defined]
+
+        # NB: prioritize unbacked symints for solving by ordering them last.
+        # Prefer to simplify out symbols created later (i.e. simplify out s4 over s3).
+        # Prefer to simplify out symbols with ephemeral sources.
+        def _smart_symbol_sort(x):
+            has_ephemeral_source = (
+                x in self.var_to_sources and any(s.is_ephemeral() for s in self.var_to_sources[x])
```

Comment:
The any test here is kind of strange. If a symbol is associated with a non-ephemeral source, then all is fine, right? I've managed to unify the symbol with an actual input source and there is no reason to use it or not. So what you're worried about is the case when all of the sources are ephemeral (but in practice, you have exactly one source and it is ephemeral).

Reply:
This also makes me wonder, if you end up with an EphemeralSource in `var_to_sources` with other stuff, you sure what to make sure you delete it later...

Reply:
good points! changed the check to `all()` and added some logic to prefer non-ephemeral sources first in the `var_to_sources` list when possible, since the first ones are potentially used later on for guarding, etc.